File: app/src/main/java/com/oliver/weatherapp/data/local/WeatherDatabase.java
package com.oliver.weatherapp.data.local;

import android.arch.persistence.room.Database;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;
import android.arch.persistence.room.TypeConverters;
import android.content.Context;
import android.util.Log;

import com.oliver.weatherapp.data.local.dao.CitiesDao;
import com.oliver.weatherapp.data.local.dao.WeatherDao;
import com.oliver.weatherapp.data.local.model.CityEntry;
import com.oliver.weatherapp.data.local.model.WeatherEntry;

@Database(entities = {CityEntry.class, WeatherEntry.class}, version = 1)
@TypeConverters(DateConverter.class)
public abstract class WeatherDatabase extends RoomDatabase {

    private static final String LOG_TAG = WeatherDatabase.class.getSimpleName();
    private static final String DATABASE_NAME = "weather";

    // For Singleton instantiation
    private static final Object LOCK = new Object();
    private static WeatherDatabase sInstance;

    public static WeatherDatabase getInstance(Context context) {
        Log.d(LOG_TAG, "Getting the database");
        if (sInstance == null) {
            synchronized (LOCK) {
                sInstance = Room
                        .databaseBuilder(context.getApplicationContext(), WeatherDatabase.class, DATABASE_NAME)
                        .build();
                Log.d(LOG_TAG, "Made new database");
            }
        }
        return sInstance;
    }

    public abstract CitiesDao citiesDao();

    public abstract WeatherDao weatherDao();
}

File: app/src/main/java/com/oliver/weatherapp/data/local/WeatherDatabase.kt
package com.oliver.weatherapp.data.local

import android.arch.persistence.room.Database
import android.arch.persistence.room.Room
import android.arch.persistence.room.RoomDatabase
import android.arch.persistence.room.TypeConverters
import android.content.Context

import com.oliver.weatherapp.data.local.dao.CitiesDao
import com.oliver.weatherapp.data.local.dao.WeatherDao
import com.oliver.weatherapp.data.local.model.CityEntry
import com.oliver.weatherapp.data.local.model.WeatherEntry
import timber.log.Timber

@Database(entities = [CityEntry::class, WeatherEntry::class], version = 1)
@TypeConverters(DateConverter::class)
abstract class WeatherDatabase : RoomDatabase() {

    abstract fun citiesDao(): CitiesDao

    abstract fun weatherDao(): WeatherDao

    companion object {
        private val DATABASE_NAME = "weather"

        // For Singleton instantiation
        private val LOCK = Any()
        @Volatile
        private var sInstance: WeatherDatabase? = null

        fun getInstance(context: Context): WeatherDatabase {
            Timber.d("Getting the database")
            return sInstance ?: synchronized(LOCK) {
                sInstance ?: buildDatabase(context).also {
                    Timber.d("Made new database")
                    sInstance = it
                }
            }
        }

        private fun buildDatabase(context: Context): WeatherDatabase {
            return Room
                    .databaseBuilder(context.applicationContext, WeatherDatabase::class.java, DATABASE_NAME)
                    .build()
        }
    }
}

